disambiguate() implemented

Yathartha22 opened the pull request:
implemented `disambiguate_symbols()` as suggested in #13431 .

- [x] Docstring needs to be added.

- [x] Test cases to be worked on.

I will add docstring in the next commit. Also, the test cases will fail due to the result returned being ambiguous.
The order of the result is not maintained, i.e. each time the function is called with same set of parameters different combination of result is obtained. I am not sure about this ambiguity @smichr if you could suggest why is this happening.

Yathartha22 committed:
disambiguate_symbols() implemented

Yathartha22 committed:
Merge branch 'master' into disambiguate

Yathartha22 commented:
I guess I figured out the reason for the ambiguity. It is caused by `free_symbol` getting different combinations of symbols every time. Consider the following case:

```
x=symbols('x', prime=True)
y=symbols('x', composite=True)

t = Tuple(x, y)
f = t.free_symbols   // f = {x, x}
for i in f:
	print(i.assumptions0)
```
In this f will be `{x, x}`, but in some case the `first x` will be prime and the `second x` composite and sometimes the opposite. The result is ambiguous.
@smichr @asmeurer  I am not sure if there is a way in sympy by we can get the result in a particular order. If so then the problem is solved for this PR.

smichr commented:
Did you try 'ordered'?

Yathartha22 commented:
Yes, I tried with `ordered` but it won't make a difference.
I tried using `f = sorted(t.free_symbols, key=t.index)`, and its working

smichr commented:
> but it won't make a difference.

`ordered` should work for Dummy symbols but it won't work for regular symbols with the same name but different assumptions and your proposal will fail, too, because regular symbols don't have the `dummy_index` (is that what you meant by `index`?) attribute.

```python
>>> def ok(x,y):
...     return list(ordered((x,y))) == list(ordered((y,x))) # true if same order returned
>>> ok(x,y)
True
>>> ok(Dummy('x'),Dummy('x'))
True
>>> ok(var('x', integer=True),var('x'))
False
```

I recommend the following:
```
>>> def ok(x,y):
...     key=lambda x:tuple(sorted(x.assumptions0.items()))
...     return list(ordered((x,y),keys=key)) == list(ordered((y,x),keys=key))
>>> ok(var('x',integer=True), var('x',prime=True))
True
```

Yathartha22 committed:
improved logic and added tests

Yathartha22 commented:
@smichr Do we need to disambiguate below types:

```
# at present

>>> disambiguate_symbols([Dummy('x')]*3)
(_x, _x, _x)

>>> disambiguate_symbols([Symbol('x'), Symbol('x')])
(x, x, x)
```
So they be evaluated to something like : (x, x_1, x_2)?

Yathartha22 committed:
docstring added

smichr commented:
> disambiguate below types

Do you mean, "should we handle lists of symbols?" I'm +0. Perhaps * args could be used so any symbols in any of the args (vs the single argument/expr) would be handled.

Yathartha22 commented:
>  I'm +0

I too think it should not handle list of symbols, as `disambiguate` is made to handle symbols that look alike but are actually different, like `Symbol('_x')` and `Dummy(x)`.

 > Perhaps * args could be used 

Do you something like this

```
>>> t = Tuple(Symbol('_x'),Symbol('_y'),Dummy('y'), Dummy('x'), Dummy('x'))
>>> disambiguate_symbols(t, _y)
(_y, _y_1)

```
I am not sure about this, because if one wants to disambiguate certain symbols, then only those symbols should be passed.

smichr commented:
OK, I looked more closely. The original intention was not that this routine should do the replacement but that it should simply return the mapping that provides the disambiguation (is that a word??). I'm not sure of the merits of returning a mapping or handling the object(s) themselves.

```
disambiguate(*BasicArgs) -> 
identify like-printing symbols in Basic object(s) 
and return Basic object with replacements; internally
handle by finding like-printing in Tuple(*BasicArgs).free_symbols

disambiguate(*syms) -> 
identify like-printing symbols in list of 
symbols and return the mapping that disambiguates them
```

Yathartha22 commented:
>  disambiguation (is that a word??)

Sure it is 😄.

> I'm not sure of the merits of returning a mapping or handling the object(s) themselves.

I am actually +1 with the idea that a mapping of symbols that disambiguate them is returned so one is able to perform operations exactly the same way when they were not disambiguated.

```
disambiguate(iter) ->
converting iter to Tuple(Basic) object(s)
identify like-printing symbols in Tuple object(s);internally
handle by finding like-printing in Tuple(...).free_symbols
and return mapped Tuple object that disambiguates them.
```
This is the way it is being implemented right now. Is the approach fine?

smichr commented:
> Is the approach fine?

How are you thinking of using the results? I gave an example of what I envisioned on the issue page. So I would imagine the routine to be something like this:

```
def disambiguate_symbols(*basic):
    iter = Tuple(*basic)
    free = list(ordered(iter.free_symbols, keys=lambda x: tuple(sorted(x.assumptions0.items()))))
    identify like-printing symbols in free and create a mapping {free_i: disambiguated_free_i} to return
```

So a dict rather than Tuple is returned and the API uses a star arg to allow 1 or more objects to be passed.

smichr commented:
The return value could also be made like cse: return mapping, Tuple wherein args have had the symbols replaced, e.g. `disambiguate(x/y, xx/y) -> {xx: x_1}, (x/y, x_1/y}` where `x` and `xx` are defined to have the same printing expressions.

smichr commented on sympy/core/symbol.py:
```diff
@@ -737,3 +739,50 @@ def traverse(symbols, frame):
+        new_iter=Tuple()
+        for i in iter:
+            new_iter += Tuple(i)
+    syms = ordered(new_iter.free_symbols, keys=new_iter.index)
```
I see that you are thinking the arguments passed are symbols so you can just use the index to make the ordering canonical, but I was thinking that expressions would be passed and *internally* the symbols would be extracted (in which case index won't work and perhaps what I suggested in the `ok` function would be better.

smichr commented:
Combined with what you have done, this is what I am thinking:

```python
def disambiguate(*args):
    args = Tuple(*args)
    mapping = {}
    for s in args.free_symbols:
        mapping.setdefault(str(s), []).append(s)
    reps = {}
    for k in mapping:
        if len(mapping[k]) == 1:
            continue
        syms = list(ordered(mapping[k],
            keys=lambda x: tuple(sorted(x.assumptions0.items()))))
        for i in range(1, len(syms)):
            reps[syms[i]] = Symbol("%s_%i" % (k, i),
                **syms[i].assumptions0)
    return reps, args.xreplace(reps)
```

Example:
```
>>> exprs = var('_x'), Dummy('x')/y
>>> exprs
(_x, _x/y)
>>> disambiguate(*exprs)
({_x: _x_1}, (_x_1, _x/y))
```

Yathartha22 commented:
@smichr sorry for the late response. I was slightly busy travelling, but I am understanding your point, I will update the PR ASAP.

Yathartha22 committed:
added star arg to allow more than one objects to be passed and
added capability to pass expressions.

Yathartha22 committed:
removed failing test

Yathartha22 commented:
@smichr have a look over the changes, I have made changes with the existing code.

> syms = list(ordered(mapping[k],
            keys=lambda x: tuple(sorted(x.assumptions0.items()))))

I guess this is not necessary.

Regarding returning `reps`, I m not sure whether we should return it, because firstly one cannot determine accurately what symbols are replaced  and secondly it would result in a messy output for larger inputs. Like for `Dummy('x'), Dummy('x')` would return `{_x: _x, _x: _x_1, _x: _x_2}`

smichr commented on sympy/core/symbol.py:
```diff
@@ -737,3 +739,47 @@ def traverse(symbols, frame):
+def disambiguate(*iter):
+    """
+    This provides a method to disambiguate symbols or expressions having
+    ambiguate symbols
```
Return a Tuple containing the passed expressions with distinct symbols (that appear the same when printed) replaced with numerically subscripted symbols.

smichr replied:
{I see now that this conflicts with your implementation. But consider when and how you are going to use this. *Please given an example* (for purpose of discussion) of what you would do with an expression that contains distinct but identically printing symbols so you can get a new expression showing them as distinct.

smichr replied:
Maybe I am thinking of the higher level routine and you are trying to write the lower level routine that it would use; the higher level might be "disambiguate" and the lower leve "_disambiguate_symbols".

Yathartha22 replied:
> Please given an example (for purpose of discussion) of what you would do with an expression that contains distinct but identically printing symbols so you can get a new expression showing them as distinct.
```
>>> x = symbols('x', prime=True)
>>> y = symbols('x', composite=True)
>>> t = x+y; t
x + x      #without disambiguate
>>>  res, = disambiguate(t); res
x + x_1
```

Yathartha22 replied:
> Maybe I am thinking of the higher level routine and you are trying to write the lower level routine that it would use

What should higher level routine do?
At present all the expressions, as well as the symbols, are disambiguated.

smichr commented on sympy/core/symbol.py:
```diff
@@ -737,3 +739,47 @@ def traverse(symbols, frame):
+            continue
+        for i in range(1, len(mapping[k])):
+            reps[mapping[k][i]] = Symbol("%s_%i"%(k, i), **mapping[k][i].assumptions0)
+    return tuple(ordered(new_iter.xreplace(reps)))
```
OK, I see you are doing the same as I intended...I got thrown off by the examples. And btw, I just ran into an expression -- though can't reproduce from memory -- where a minpoly had two "x" values in the result, but they we actually different variables.

A fewcomments:
1. if you don't sort then the numbering will be arbitrary and not canonical; the demo of the `ok` function shows an example wherein the (arbitrary) ordering of the symbols doesn't lead to a canonical result.
2. you definitely don't want to sort the result coming back -- someone will want to see expr[0] still in slot 0 with its new symbols; I suspect that's why you thought ordering symbols (point previous) was not necessary.
3. perhaps you could make the mapping (potentially inverted) be an optional argument; with it, you can replace the symbols with whatever you want once the routine has figured out which symbols might need to be tweaked. How could it be made a defaultdict that gives back the key if a value for that key is missing (other than doing the obvious as a user: `result = mapping.get(x, x)`)...[googling](https://stackoverflow.com/questions/10524142/python-dictionary-return-requested-key-if-value-does-not-exist):
```
from collections import defaultdict

class KeyAwareDefaultDict(defaultdict):
    def __missing__(self, key):
        if self.default_factory is None:
            raise KeyError(key)
        self[key] = value = self.default_factory(key)
        return value

>>> lookup = KeyAwareDefaultDict(Id)
>>> lookup[x]
x
>>> lookup[x] = y; lookup[x]
y
```

Yathartha22 replied:
For point 1: Yes to make it canonical, I have `ordered` the free_symbols at the start.

For point 2: Actually, an ordered tuple is returned as the answer, and this changes the ordering as it was in the input. I just did to make the output more simpler to read. If you suggest I could remove ordered in the last return statement.

For point 3: We can use as you suggest, but I am not sure whether it would make any difference. Means, it would reduce the loops but will it affect any performance of the function?, otherwise it is just another way to implement the function.(*I will try to implement it locally and check)

smichr commented:
> Yes to make it canonical, I have ordered the free_symbols at the start

See the line I marked with "this fails so simple ordering is not sufficient" above.

Yathartha22 committed:
added `key` for canonical ordering

Yathartha22 committed:
improved docstrings and added few examples

Yathartha22 commented:
I have added `key` as suggested have a look.

smichr commented on sympy/core/symbol.py:
```diff
@@ -737,3 +739,53 @@ def traverse(symbols, frame):
+    for k in mapping:
+        reps[mapping[k][0]] = Symbol("%s"%(k), **mapping[k][0].assumptions0)
+        if len(mapping[k]) == 1:
+            continue
```
I would move that line 786 to here before the "continue" then run the loop below "for i, s in enumnerate(mapping[k]): reps[s] = Symbol(...**s.assumptions0)` (prettier, however :-))

Yathartha22 replied:
Agreed, but the reason I placed there was it would output something like
`(_x_1, _x_0)`, I don't think we should have `0` subscripted elements(doesn't looks nice 😄). What do you say?

smichr commented on sympy/core/symbol.py:
```diff
@@ -737,3 +739,53 @@ def traverse(symbols, frame):
+def disambiguate(*iter):
+    """
+    Return a Tuple containing the passed expressions with distinct
+    symbols(that appear the same when printed) replaced with numerically
```
space before "(" and it's worth noting that all free symbols that were Dummies will now be Symbols.

Yathartha22 replied:
> it's worth noting that all free symbols that were Dummies will now be Symbols.

Yes, I noticed it earlier and thought that rather than converting everything to `Symbol`, the conversion should be according to the class they belong. So I tried something like this:
`reps[mapping[k][i]] = mapping[k][i].__class__("%s_%i"%(k, i), **mapping[k][i].assumptions0)`, but  this created a slight problem in creating tests for `Dummy` items(two identical dummies are not equal), therefore I thought it would be better if everything is `Symbol`.

smichr committed:
update docstring

Update docstring and remove continue: let loop handle case of len(mapping[k]) == 1

smichr committed:
only include k0 if it's different than k

smichr commented:
> I don't think we should have 0 subscripted element

I see, now, why that line has to be there and I agree. I tweaked things a little. I'm happy with it. Do you have any other changes you would like to see?

Yathartha22 commented:
> Do you have any other changes you would like to see?

LGTM +1. I like the example that you have added.
As of now everythig looks good.

smichr committed:
remove use of var in test

smichr commented:
OK...if the last mods pass I will commit this unless anybody else has comments.

smichr committed:
remove unmatched parenthesis

smichr commented:
What do you think about using str(k).strip('_') as the name to test -- that would get rid of the annoying underscore prefix and suffix and create symbols that will actually look right when printed. I wonder if anythign can be done about 'x_1' becoming 'x_1_1' and looking like 'x_11' when printed. Maybe we could posfix a '_,1'.

smichr committed:
correct t7 test

smichr commented on sympy/core/tests/test_symbol.py:
```diff
@@ -366,3 +367,24 @@ def test__uniquely_named_symbol_and__symbol():
+    t2 = Dummy('x'), Dummy('x')
+    t3 = Dummy('x'), Dummy('y')
+    t4 = x, Dummy('x')
+    t5 = Symbol('x', integer=True), x
```
If we add `Symbol('x_1') to `t5` I think we will run into problems: the index of the duplicate symbol cannot be added blindly without checking to see if the proposed symbol name is already in the list of known symbols. This might also be a reason to strip underscores *and* numbers (like `'_x_2_3_'.strip('_1234567890') -> x`).

smichr committed:
strip '_' from prefix; watch for clashing symbols

smichr committed:
Update test_symbol.py

smichr commented:
ping @Yathartha22  -- how do the changes look to you?

Yathartha22 commented:
@smichr I am really sorry that I have not looked into the changes because I have been busy for these couple of days. I am +1 with the idea of stripping the underscores, but I will look into the changes that you made thoroghly and get back to you in the morning, if its okay for you.

smichr commented:
Just happened to see your post. No hurry.

smichr committed:
Update symbol.py

smichr committed:
Update test_symbol.py

Yathartha22 commented:
@smichr the changes looks great to me. I guess the disparities are solved with these changes +1

Yathartha22 commented on sympy/core/tests/test_symbol.py:
```diff
@@ -366,3 +367,27 @@ def test__uniquely_named_symbol_and__symbol():
+
+
+def test_disambiguate():
+    x, y, y_1, _x, _x_1, x_1, x_2 = symbols('x y y_1 _x _x_1 x_1 x_2')
```
I guess `_x_1` must be removed if it is not used. If you say so I will remove it.

smichr committed:
Update test_symbol.py

smichr commented:
Got it. Ready to go when tests pass. Thanks for taking this up.
